server/src/test/java/org/elasticsearch/index/mapper/XContentDataHelperTests.java

```diff
@@ -0,0 +1,107 @@
+import org.elasticsearch.xcontent.json.JsonXContent;
+
+import java.io.IOException;
+import java.math.BigDecimal;
+import java.math.BigInteger;
+import java.util.List;
+
+import static org.hamcrest.Matchers.equalTo;
+
+public class XContentDataHelperTests extends ESTestCase {
+
+    private String encodeAndDecode(String value) throws IOException {
+        XContentParser p = createParser(JsonXContent.jsonXContent, "{ \"foo\": " + value + " }");
+        assertThat(p.nextToken(), equalTo(XContentParser.Token.START_OBJECT));
+        assertThat(p.nextToken(), equalTo(XContentParser.Token.FIELD_NAME));
+        assertThat(p.currentName(), equalTo("foo"));
+        p.nextToken();
+
+        XContentBuilder builder = XContentFactory.jsonBuilder();
+        builder.humanReadable(true);
+        XContentDataHelper.decodeAndWrite(builder, XContentDataHelper.encodeToken(p));
+        return Strings.toString(builder);
+    }
+
+    public void testBoolean() throws IOException {
+        boolean b = randomBoolean();
+        assertEquals(b, Boolean.parseBoolean(encodeAndDecode(Boolean.toString(b))));
+    }
+
+    public void testString() throws IOException {
+        String s = "\"" + randomAlphaOfLength(5) + "\"";
+        assertEquals(s, encodeAndDecode(s));
+    }
+
+    public void testInt() throws IOException {
+        int i = randomInt();
+        assertEquals(i, Integer.parseInt(encodeAndDecode(Integer.toString(i))));
+    }
+
+    public void testLong() throws IOException {
+        long l = randomLong();
+        assertEquals(l, Long.parseLong(encodeAndDecode(Long.toString(l))));
+    }
+
+    public void testFloat() throws IOException {
+        float f = randomFloat();
+        assertEquals(0, Float.compare(f, Float.parseFloat(encodeAndDecode(Float.toString(f)))));
+    }
+
+    public void testDouble() throws IOException {
+        double d = randomDouble();
+        assertEquals(0, Double.compare(d, Double.parseDouble(encodeAndDecode(Double.toString(d)))));
+    }
+
+    public void testBigInteger() throws IOException {
+        BigInteger i = randomBigInteger();
+        assertEquals(i, new BigInteger(encodeAndDecode(i.toString()), 10));
+    }
+
+    public void testBigDecimal() throws IOException {
+        BigDecimal d = new BigDecimal(randomBigInteger(), randomInt());
+        XContentBuilder builder = XContentFactory.jsonBuilder();
+        builder.startObject().field("foo", d).endObject();
+        XContentParser p = JsonXContent.jsonXContent.createParser(parserConfig(), BytesReference.bytes(builder).streamInput());
+        assertThat(p.nextToken(), equalTo(XContentParser.Token.START_OBJECT));
+        assertThat(p.nextToken(), equalTo(XContentParser.Token.FIELD_NAME));
+        assertThat(p.currentName(), equalTo("foo"));
+        p.nextToken();
+        assertEquals(p.numberType(), XContentParser.NumberType.DOUBLE);
+
+        // FieldDataParseHelper.decodeAndWrite(builder, FieldDataParseHelper.encodeToken(p));
```

Comment:
Should this be uncommented?

Reply:
This is actually not working, thanks for reminding me. I've no good way to test BigDecimal, it gets converted to double under the covers.. `IgnoreMalformedStoredValuesTests` does test this so I'm not terribly worried, wdyt?

Reply:
let me check this

Reply:
Yes, I think it is fine.